### ISSUE TITLE
Importing into CSS Module

### ISSUE THREAD

**Issue by willmcclellan** [css-modules/css-modules#22]
I'm wondering what method people are using for importing data into their CSS Modules. `composes` is great for extending/composing styles but what about data such as color variables or measurement defaults?

I had a look at the ICSS spec and the `:import` syntax (https://github.com/css-modules/icss#import) but it doesn't look suited to for this use case.

Are there any patterns/techniques other people are using for this?

**Comment by nkbt** [comment]
Why not do class names composition in the component instead and not add more side-effects to the css?

**Comment by andreypopp** [comment]
I'm also interested in this.

@nkbt not sure what side-effectful in sharing constants between modules

**Comment by nkbt** [comment]
css file includes something else from outside. It is not self-contained and pure anymore.

See this comment https://github.com/css-modules/css-modules/issues/21#issuecomment-121429234

I believe this is the best approach.

**Comment by geelen** [comment]
Yeah, so ICSS *could* support arbitrary symbols, not just class names, which could be used for variables. The first thing we're going to use that for is for media queries to have first-class treatment in CSS Modules. I think we're going to lean on [postcss-custom-media](https://github.com/postcss/postcss-custom-media) to do something like the following:

```css
/* breakpoints.css (media queries are exported by default) */
@custom-media --small (max-width: 30em);
@custom-media --med (max-width: 60em);
@custom-media --large (min-width: 60.01em);
```

```css
/* my-component.css */
@custom-media --small from "./breakpoints.css";

@media (--small) {
  /* styles for small viewport */
}
```

This isn't in place yet, but I'm working towards it.

Now, for variables, we could do something using [postcss-custom-properties](https://github.com/postcss/postcss-custom-properties):

```css
/* constants.css */
:root {
  --primary: rgb(248, 10, 10);
}
```

```css
/* my-component.css - option 1 */
:root {
  --primary: primary from "./constants.css";
}

.foo {
  color: var(--primary);
}
```

```css
/* my-component.css - option 2 */
.foo {
  color: var(--primary from "./constants.css");
}
```

But I'm not sure I like that so much. For one, the postcss-custom-properties is like a polyfill, but it's limited in what it can do (because the spec has runtime behaviour that can't be polyfilled). As such, I'm not sure how valuable it is to be using a cut-down version of the spec. But the bigger issue is the fact that I don't see a strong use case for variables *at all* with CSS Modules. Basically, with variables, you import a number, or a string, or a colour. With CSS Modules you can **compose** a class with the same number, string or colour **and the context** in which it's used. So, instead of:

```css
.foo {
  color: var(--primary);
  border-radius: var(--small-border-radius);
}
```

you can just write:

```css
.foo {
  composes: primary-color from "./colors.css";
  composes: slightly-rounded from "./borders.css";
}
```

So, in summary, special media query syntax :+1:, variable support... probably not.

**Comment by nkbt** [comment]
I don't quite understand why add any kind of logic to css at all. We already have JS and can manipulate what and how we import and compose.

**Comment by geelen** [comment]
That's what I'm arguing for. Except there is no "composition" for media queries, so we add special syntax there.

**Comment by willmcclellan** [comment]
Thanks for the input, it certainly helps. Currently we're using a post-css plugin but it's global so kind of defies the point of modules.

I did experiment with `composes` but it's a little bit restrictive, in that to achieve this:

````
.foo {
  color: var(--primary);
  background-color: var(--primary);
  border-radius: var(--small-border-radius);
}
````

You would need this:

````
.foo {
  composes: primary-color from "./colors.css";
  composes: primary-background-color from "./colors.css";
  composes: slightly-rounded from "./borders.css";
}
````

That's a simple example that's starting to get a little convoluted, because the variable is tightly coupled to the style property. Maybe I'm thinking about it in the wrong way, but from a JS perspective this kind of thing would be trivial:

````
import colors from './colors'
import borders from './borders'

.foo = {
  color: colors.primary,
  background: colors.primary,
  borderRadius: borders.slightlyRounded
}
````

But yeh, this approach might not be ideal altogether because it makes the CSS too dependant. Perhaps we should be composing the themes-based styles outside of the CSS and passing them into components.

**Comment by geelen** [comment]
You can compose multiple classes on one line, which makes things a bit more descriptive if you pick good names:

```css
.foo {
  composes: primary on-light from "./colors.css";
}
```

Here's another example, with just some names off the top of my head:

```css
.bar {
  composes: alternate on-dark from "./colors.css";
  composes: light-rounded-border light-box-shadow-on-hover from "./borders.css";
  composes: tooltip top-right from "./layout.css";
  composes: fades-in-on-hover from "./effects.css";
}
```

**Comment by willmcclellan** [comment]
Ah, that's really nice. :+1:

**Comment by nkbt** [comment]
There was a nice talk on React Europe
https://www.youtube.com/watch?v=ERB1TJBn32c

Separating your styles on visual and "state"-styles works pretty well for
us. Visual are coming from css-modules, state-styles are done via js.

**Comment by geelen** [comment]
Gonna close this just to tidy up our issues. Feel free to reopen if you have further questions, @willmcclellan :)

**Comment by andreypopp** [comment]
I implemented an API to generate CSS modules out of JS which solves that problem. https://github.com/andreypopp/styling

It has a webpack loader which executes JS code during building and generates CSS modules from exported values. You can use JS to construct styling objects.

**Comment by serapath** [comment]
I was giving an example that has not much to do with the way `css-modules` work, but it gets the job done. See https://github.com/css-modules/css-modules/issues/23

In `ICSS` it says [here](https://github.com/css-modules/icss#specification)
```css
:import("path/to/dep.css") {
  localAlias: keyFromDep;
  /* ... */
}
:export {
  exportedKey: exportedValue;
    /* ... */
}
```
Maybe it's totally off, but what about something like:
```css
:root {
  --myThemeColor1: blue;
}

 /* as in npm modules, but with the option to pass in some customizations */
:import('moduleName')(var(--myThemeColor1)) {
  localAlias: keyFromDep;
}
:export {
  exportedKey: exportedValue;
    /* ... */
}
```
Because it's nice to be able to import css and to choose what is needed, but how to adapt a certain `css component` with some styling that is custom to a specific project that is being worked on?
It would be a shame if a `css module` would define some kind of `padding` or `font-size` or `color` that cannot be switched to values needed in the current project.

Of course, once it's loaded into the project, I can always use `javascript` to adapt a specific styling to what i need in a given project, but it would be nice if the initial css can already be adapted.

Beyond that, it would even be nice, if there could be some kind of `live updates`
```javascript
import styles from './App.css';
/* .... */
render() {
    return (
      <div className={styles.app}>
/* ... */
/* probably there is a nicer way to specify the API for that, but maybe something like: */
styles(styles.app).update('color', '#f00')
```

**Comment by F21** [comment]
This is the solution I am using with `postcss`. I installed `postcss-cssnext` and `postcss-import`.

The relevant bits in my webpack config (`postcss-import` must be the first plugin):

```javascript
postcss: function() {
    return [
        require('postcss-import')(),
        require('postcss-cssnext')(),
    ];
}
```

Then, I have a `variables.css` file:

```css
:root{
    --color: #666666;
}
```

Then, in my `component.css` file which is being used by a react component:

```css
@import 'path/to/variables.css';

.some-component-container{
    border-bottom: 1px solid var(--secondary-color);
}
```

**Comment by SH4DY** [comment]
@F21 This is the cleanest solution. Exactly what I was looking for! This would be a great fit for the README of the postcss-import or postcss-next repos.

**Comment by caesarsol** [comment]
@F21 I'm trying with your approach, but it doesn't work! 

```
WARNING in ./~/css-loader?modules&importLoaders=1!./~/postcss-loader!./components/Checkbox/style.css
postcss-custom-properties: ./components/Checkbox/style.css:5:3: variable '--active-color' is undefined and used without a fallback
```

Any suggestions? :)

**Comment by F21** [comment]
@caesarsol I'm not sure. It's been a while since I worked with the front-end, so I haven't updated my webpack, postcss, etc to the latest versions. Can you post the relevant parts of your webpack config as well as the css you are using?

**Comment by caesarsol** [comment]
Hey @F21 thanks! 
My webpack config is as follows:

```js
const path = require('path');

module.exports = {
  module: {
    loaders: [
      {
        test: /\.css?$/,
        loader: 'style!css-loader?modules&importLoaders=1&localIdentName=[name]__[local]__[hash:base64:5]!postcss-loader'
      }
    ]
  },
  postcss: function () {
    return [
      require('postcss-import'),
      require('postcss-cssnext'),
    ];
  }
}
```

And the CSS files:

```css
/* variables.css */
:root {
  --primary-color: tomato;
  --active-color: tomato;
  --inactive-color: lightgrey;
}
```

```css
/* component.css */
@import 'variables.css';

.checked {
  background-color: var(--active-color);
}
```

**Comment by F21** [comment]
@caesarsol I just made a sample repo: https://github.com/F21/css-modules-import

I think your problem is because postcss wasn't able to import the variables file. In that case, you need to set a folder to start searching for css files in `postcss-import`. In my example, I used `require('postcss-import')({path: __dirname+"/src"}),`.

**Comment by raspo** [comment]
@caesarsol I was having the same issue.
Removing `&importLoaders=1` from the loader configuration fixed it for me.

```
{
  test: /\.css$/,
  loader: 'style-loader!css-loader?modules&sourceMap!postcss-loader'
}
```

**Comment by marcselman** [comment]
I'm getting the same errors when starting node
`variable '--primary-color' is undefined and used without a fallback`
But the strange thing is that it does work. I can't seem to get rid of the error but the variables work just fine...

**Comment by F21** [comment]
@marcselman I am not getting that error. See https://github.com/F21/css-modules-import

Just run `npm install` and `npm run build`.

I suspect the reason you are seeing that error is due to the ordering of your PostCSS plugins. Do you have anything else besides `postcss-import` and `postcss-next`?

**Comment by marcselman** [comment]
I found out what the problem is. I'm using the [css-modules-require-hook](https://github.com/css-modules/css-modules-require-hook) to compile css-modules at runtime. But that module uses require which is synchronous and I can't add postcss-import to the hook because it's asynchronous. That's why I receive the undefined variable errors because the import statements are not first inlined when the css modules get compiled.

I have not found a simple solution for this so for now I've chosen to only use composes statements instead of import statements.

**Comment by techyrajeev** [comment]
@F21  Your solution seems to be elegant. I just want to know whether is there any way to pass color variable from javascript( or jsx). e.g suppose I have a button component and I want to set the color value of button dynamically ( may be after getting value from server side).

**Comment by joshwnj** [comment]
@techyrajeev you can pass values from js to css with a postcss plugin like https://github.com/macropodhq/postcss-constants or https://github.com/MadLittleMods/postcss-css-variables

**Comment by techyrajeev** [comment]
@joshgillies Thanks for the info. Can I change the css generated at runtime with this after passing variable value fetched from server and apply to some component?

**Comment by joshwnj** [comment]
@techyrajeev with this many moving parts it might be helpful to set up a simple example repo that demonstrates what you want. There's a few different ways of going about it so I'm sure someone here will be able to offer advice on how to make it work.

**Comment by everdimension** [comment]
I am currently using this technique to import variables into css-modules files:

```css
/* styles/colors.css */
@value brandColor: #abc;
@value borderWidth: 2px;
@value textColor: #222;
```

In a component file:

```css
/* components/Button.css */
@value brandColor, borderColor from '../styles/colors.scss';

:root {
  --brandColor: brandColor;
  --borderWidth: borderWidth;
}

.Button {
  /* ... button styles ... */
  background-color: var(--brandColor);
  border: var(--borderWidth) solid;
}
```

It involves some repetition but I think it's very clear and can be clearly rewritten into something else if you decide to abandon it.

**Comment by silasb** [comment]
@everdimension why do you wrap it in

```css
:root {
  ...
}
```

instead of just doing 

```css
.Button {
   background-color: brandColor;
}
```

I'm a newb when it comes to css modules.

**Comment by everdimension** [comment]
@silasb because that's the `cssnext` [custom properties syntax](http://cssnext.io/features/#custom-properties-var)
You don't have to use it and you can do just like you proposed.

**Comment by danny-andrews** [comment]
> ```
> .foo {
>   composes: primary on-light from "./colors.css";
> }
> ```

@geelen That example works fine until you want to use the same color value as a `background-color` in one place and a `color` in another.

**Comment by Download** [comment]
...which would look like:

```css
.foo {
  composes: light on_primary
}
```

It is just a matter of choosing smart primitives to compose and naming them in a smart way. `on-light` and `on-primary` clearly indicate this is about the background color. And `primary on-light` and `light on-primary` are very readable imho.

The problem for me is that the `composes` keyword immediately 'downgrades' the file from 'universal CSS' to 'CSS Modules CSS'.

I love the fact that I can take a normal CSS file (or even a SASS file) and turn it into a CSS Module. Using `composes` works only if I turn it into a CSS Module. Making CSS Modules mandatory. I prefer to keep my CSS as 'vanilla' CSS so it *can* be used as a CSS Module but does not have to.

I use SASS and SASS variables i.c.w. CSS Modules. Just set variables in SASS like so:

*variables.scss*
```css
$primary: green !default;
```

Then use it in your stylesheet like so:

*coolstyles.scss*
```css
@import 'variables.scss';

.foo {
  background: $primary;
}
```

then, in your component CSS, you can just set the variable yourself *before* importing the *coolstyles.scss*. This works because the `!default` part of the rule basically says 'set this var only if it does not have a value yet'. So when you set the variable first, it wins. As an example:

*style.scss*
```css
$primary: blue;
@import 'coolstyles.scss';
```

*MyComponent.jsx*
```js
import style from './style.scss';

export const MyComponent = (props) => (
  <div className={style.foo}>Hello, World!</div>
)
```

MyComponent will end up rendering something like:

```html
<div class="foo_2Z6gf">Hello, World!</div>
```

..and the CSS will look something like:

```css
.foo_2z6gf {
  background: blue;
}
```

Any references to green are completely gone, because the SASS logic runs before the CSS Modules logic.

So in short, I simply use SASS variables (and mixins etc.) in combination with CSS Modules and that works great!
